File: packages/unswallow/src/validate.ts
import type { ToolEnvelope, ToolSchema } from './types';

export interface ToolValidationResult {
  /** The recovered call has the minimum shape required by OpenAI-compatible tool calls. */
  structurallyValid: boolean;
  /** Whether the recovered name appears in the supplied tool schemas. */
  nameKnown: 'yes' | 'no' | 'unknown';
  /** Whether arguments satisfy the supplied tool's supported JSON Schema subset. */
  schemaValid: 'yes' | 'no' | 'unknown';
  errors: string[];
}

type JsonObject = Record<string, unknown>;

function isObject(value: unknown): value is JsonObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function schemaFor(tool: ToolSchema): JsonObject | null {
  const parameters = tool.function?.parameters ?? tool.parameters;
  return isObject(parameters) ? parameters : null;
}

function nameFor(tool: ToolSchema): string | undefined {
  const name = tool.function?.name ?? tool.name;
  return typeof name === 'string' && name.trim() ? name.trim() : undefined;
}

function typeMatches(value: unknown, type: string): boolean {
  switch (type) {
    case 'object': return isObject(value);
    case 'array': return Array.isArray(value);
    case 'string': return typeof value === 'string';
    case 'number': return typeof value === 'number' && Number.isFinite(value);
    case 'integer': return typeof value === 'number' && Number.isInteger(value);
    case 'boolean': return typeof value === 'boolean';
    case 'null': return value === null;
    default: return true;
  }
}

function validateValue(value: unknown, schema: JsonObject, path: string, errors: string[]): void {
  const type = schema.type;
  if (typeof type === 'string' && !typeMatches(value, type)) {
    errors.push(`${path} must be ${type}`);
    return;
  }
  if (Array.isArray(schema.enum) && !schema.enum.some((candidate) => Object.is(candidate, value))) {
    errors.push(`${path} must be one of the allowed values`);
  }
  if ('const' in schema && !Object.is(schema.const, value)) errors.push(`${path} must equal the required value`);

  if (isObject(value)) {
    const properties = schema.properties;
    if (properties !== undefined && !isObject(properties)) {
      errors.push(`${path} schema has non-object properties`);
      return;
    }
    if (Array.isArray(schema.required)) {
      for (const key of schema.required) {
        if (typeof key === 'string' && !(key in value)) errors.push(`${path}.${key} is required`);
      }
    }
    if (isObject(properties)) {
      for (const [key, childSchema] of Object.entries(properties)) {
        if (key in value && isObject(childSchema)) validateValue(value[key], childSchema, `${path}.${key}`, errors);
      }
      if (schema.additionalProperties === false) {
        for (const key of Object.keys(value)) {
          if (!(key in properties)) errors.push(`${path}.${key} is not allowed`);
        }
      }
    }
  }

  if (Array.isArray(value) && isObject(schema.items)) {
    value.forEach((item, index) => validateValue(item, schema.items as JsonObject, `${path}[${index}]`, errors));
  }
}

/**
 * Validate a recovered tool envelope without attempting to repair it. Only a
 * conservative JSON Schema subset is evaluated; unsupported or malformed
 * schemas produce an `unknown` result rather than a false validation claim.
 */
export function validateEnvelope(
  envelope: Pick<ToolEnvelope, 'name' | 'arguments'> | null | undefined,
  toolSchemas?: ToolSchema[]
): ToolValidationResult {
  if (!envelope || typeof envelope.name !== 'string' || !envelope.name.trim() || !isObject(envelope.arguments)) {
    return { structurallyValid: false, nameKnown: 'unknown', schemaValid: 'unknown', errors: ['tool envelope must have a name and object arguments'] };
  }
  if (!toolSchemas || toolSchemas.length === 0) {
    return { structurallyValid: true, nameKnown: 'unknown', schemaValid: 'unknown', errors: [] };
  }
  const matching = toolSchemas.filter((tool) => nameFor(tool) === envelope.name);
  if (matching.length === 0) {
    return { structurallyValid: true, nameKnown: 'no', schemaValid: 'unknown', errors: [`recovered tool name "${envelope.name}" not found in provided toolSchemas`] };
  }
  const schema = schemaFor(matching[0]);
  if (!schema) {
    return { structurallyValid: true, nameKnown: 'yes', schemaValid: 'unknown', errors: [`tool schema for "${envelope.name}" has no object parameters schema`] };
  }
  if (schema.properties !== undefined && !isObject(schema.properties)) {
    return { structurallyValid: true, nameKnown: 'yes', schemaValid: 'unknown', errors: [`tool schema for "${envelope.name}" has non-object properties`] };
  }
  if (schema.required !== undefined && !Array.isArray(schema.required)) {
    return { structurallyValid: true, nameKnown: 'yes', schemaValid: 'unknown', errors: [`tool schema for "${envelope.name}" has non-array required`] };
  }
  const errors: string[] = [];
  validateValue(envelope.arguments, schema, 'arguments', errors);
  return { structurallyValid: true, nameKnown: 'yes', schemaValid: errors.length === 0 ? 'yes' : 'no', errors };
}

File: packages/python/unswallow/validate.py
from __future__ import annotations

import math
from typing import Any, Dict, List, Optional, Union

from .types import ToolEnvelope, ToolValidationResult


def _is_object(value: Any) -> bool:
    return isinstance(value, dict)


def _schema_for(tool: Any) -> Optional[Dict[str, Any]]:
    if not _is_object(tool):
        return None
    function = tool.get("function")
    container = function if _is_object(function) else tool
    parameters = container.get("parameters")
    return parameters if _is_object(parameters) else None


def _name_for(tool: Any) -> Optional[str]:
    if not _is_object(tool):
        return None
    function = tool.get("function")
    container = function if _is_object(function) else tool
    name = container.get("name")
    return name.strip() if isinstance(name, str) and name.strip() else None


def _type_matches(value: Any, expected: str) -> bool:
    if expected == "object":
        return _is_object(value)
    if expected == "array":
        return isinstance(value, list)
    if expected == "string":
        return isinstance(value, str)
    if expected == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool) and math.isfinite(value)
    if expected == "integer":
        return (
            isinstance(value, int)
            and not isinstance(value, bool)
            or isinstance(value, float)
            and math.isfinite(value)
            and value.is_integer()
        )
    if expected == "boolean":
        return isinstance(value, bool)
    if expected == "null":
        return value is None
    return True


def _same_json_value(left: Any, right: Any) -> bool:
    if (
        isinstance(left, (int, float))
        and not isinstance(left, bool)
        and isinstance(right, (int, float))
        and not isinstance(right, bool)
    ):
        return left == right or (
            isinstance(left, float)
            and isinstance(right, float)
            and math.isnan(left)
            and math.isnan(right)
        )
    return type(left) is type(right) and left == right


def _validate_value(value: Any, schema: Dict[str, Any], path: str, errors: List[str]) -> None:
    expected_type = schema.get("type")
    if isinstance(expected_type, str) and not _type_matches(value, expected_type):
        errors.append("{} must be {}".format(path, expected_type))
        return

    enum = schema.get("enum")
    if isinstance(enum, list) and not any(_same_json_value(candidate, value) for candidate in enum):
        errors.append("{} must be one of the allowed values".format(path))
    if "const" in schema and not _same_json_value(schema["const"], value):
        errors.append("{} must equal the required value".format(path))

    if _is_object(value):
        properties = schema.get("properties")
        if properties is not None and not _is_object(properties):
            errors.append("{} schema has non-object properties".format(path))
            return
        required = schema.get("required")
        if isinstance(required, list):
            for key in required:
                if isinstance(key, str) and key not in value:
                    errors.append("{}.{} is required".format(path, key))
        if isinstance(properties, dict):
            for key, child_schema in properties.items():
                if key in value and isinstance(child_schema, dict):
                    _validate_value(value[key], child_schema, "{}.{}".format(path, key), errors)
            if schema.get("additionalProperties") is False:
                for key in value:
                    if key not in properties:
                        errors.append("{}.{} is not allowed".format(path, key))

    items = schema.get("items")
    if isinstance(value, list) and isinstance(items, dict):
        for index, item in enumerate(value):
            _validate_value(item, items, "{}[{}]".format(path, index), errors)


def validate_envelope(
    envelope: Optional[Union[ToolEnvelope, Dict[str, Any]]],
    tool_schemas: Optional[List[Dict[str, Any]]] = None
) -> ToolValidationResult:
    """Validate a recovered envelope against a conservative JSON Schema subset.

    Malformed schemas are reported as unknown rather than as a validation
    failure, so callers never receive an unjustified validation claim.
    """
    if isinstance(envelope, dict):
        name = envelope.get("name")
        arguments = envelope.get("arguments")
    else:
        name = getattr(envelope, "name", None) if envelope is not None else None
        arguments = getattr(envelope, "arguments", None) if envelope is not None else None
    if not isinstance(name, str) or not name.strip() or not _is_object(arguments):
        return ToolValidationResult(
            False,
            "unknown",
            "unknown",
            ["tool envelope must have a name and object arguments"],
        )
    if not tool_schemas:
        return ToolValidationResult(True, "unknown", "unknown", [])

    matching = [tool for tool in tool_schemas if _name_for(tool) == name]
    if not matching:
        return ToolValidationResult(
            True,
            "no",
            "unknown",
            ['recovered tool name "{}" not found in provided toolSchemas'.format(name)],
        )

    schema = _schema_for(matching[0])
    if schema is None:
        return ToolValidationResult(
            True,
            "yes",
            "unknown",
            ['tool schema for "{}" has no object parameters schema'.format(name)],
        )
    if "properties" in schema and not _is_object(schema["properties"]):
        return ToolValidationResult(
            True,
            "yes",
            "unknown",
            ['tool schema for "{}" has non-object properties'.format(name)],
        )
    if "required" in schema and not isinstance(schema["required"], list):
        return ToolValidationResult(
            True,
            "yes",
            "unknown",
            ['tool schema for "{}" has non-array required'.format(name)],
        )

    errors: List[str] = []
    _validate_value(arguments, schema, "arguments", errors)
    return ToolValidationResult(True, "yes", "yes" if not errors else "no", errors)
